### ISSUE TITLE
[TBC] Unit::IsFriendlyTo(Unit const* unit) const bug

### ISSUE THREAD

**Issue by Ve1vet** [cmangos/issues#1160]
**Current behavior**: 
I spend all night trying to find out wtf...
Function `bool Unit::IsFriendlyTo(Unit const* unit) const`
This check:
```cpp
// always non-friendly to enemy
    if (getVictim() == unit || unit->getVictim() == this)
        return false;

```
Destroy spells  with damage that have this targets(maybe other):
Targets (22, 30) (TARGET_CASTER_COORDINATES, TARGET_ALL_FRIENDLY_UNITS_AROUND_CASTER)

**Expected behavior**: 
For example, spell Fatal Attraction. Spell 41001 apply aura on player, that trigger spell 40870 every second(30 sec aura). Spell 40870 in its turn triggers 40871, that sould deal damage to all friendly units in range (radius 25).
Both 40870 & 40871 have:
Targets (22, 30) (TARGET_CASTER_COORDINATES, TARGET_ALL_FRIENDLY_UNITS_AROUND_CASTER)

Both spells should have in targetmap all friendly units in rane to hit them wit AOE.

**Steps to reproduce**:

1.  In SpellEffects.cpp in void Spell::EffectDummy(SpellEffectIndex eff_idx) for SPELLFAMILY_GENERIC case add:
```cpp
case 40870:                                 // Fatal Attraction
{
    if(!unitTarget || unitTarget->GetTypeId() != TYPEID_PLAYER )
        return;               
    m_caster->CastSpell(m_caster, 40871, true);
        return;
}
```
2. Compile core.
3. Start server.
4. Get 2 friendly characters in range 25.
5. For example, cast on one of them 41001.
6. Spell 40871 will hit you only once. Than friendly target become unfiendly by this check:
` Unit* getVictim() const { return m_attacking; }`
7. Victim target gets permanent combat non visible. On logout charcter dies.

**Client version**: 
2.4.3
**Revision**: 
Latest

P.S. In spell.cpp in case TARGET_ALL_FRIENDLY_UNITS_AROUND_CASTER:
After filling area targets in case cpell 40870, if targetUnitMap.empty() we should remove 41001 aura from caster.

**Comment by vovk** [comment]
I change:
```
// always non-friendly to enemy
    if (getVictim() == unit || unit->getVictim() == this)
        return false;
```
to(little hack for test):
```
// always non-friendly to enemy
    if (getVictim() == unit || unit->getVictim() == this) {
        Unit const* thisOwner = GetCharmerOrOwner();
        Unit const* unitOwner = unit->GetCharmerOrOwner();
        Unit const* rThis = thisOwner ? thisOwner : this;
        Unit const* rOwner = unitOwner ? unitOwner : unit;
        if (rThis->GetTypeId() == TYPEID_PLAYER && rOwner->GetTypeId() == TYPEID_PLAYER) {
            Player const* pThis = (Player const*)rThis;
            Player const* pTarget = (Player const*)rOwner;

            // Duel
            if (pThis->IsInDuelWith(pTarget))
                return false;

            // Group
            if (pThis->GetGroup() == pTarget->GetGroup())
                return true;

        }
        return false;
    }
```
Spells 40870 & 40871 starts work like it should. So function `bool Unit::IsFriendlyTo(Unit const* unit) const` should be reworked properly.
More info:
1. Character still dies after logout.
2. Pet(in my case hunter) when takes damage from 40871 agroes on friendly caster & start hit him.

**Comment by Warlockbugs** [comment]
This is another case of friendly fire issue. This should be fixed inside damage dealing.

**Comment by AnonXS** [comment]
Needs to be rechecked, API related to this was completely reworked.

**Comment by BenDol** [comment]
Little hard to follow this issue, I tried to reproduce as suggested but wasn't really able to figure it out. Is this still an issue?

**Comment by Warlockbugs** [comment]
No, because i replaced entire underlying API with an authentic reverse engineered one.
